packages/override-tools/src/ActionQueue.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ *
+ * @format
+ */
+
+/**
+ * Executes actions sequentially, fullfilling the returned promise once complete.
+ */
+export default class ActionQueue<T> {
+  private actions: Array<() => Promise<void>> = [];
+
+  enqueue(action: () => Promise<T>): Promise<T> {
+    return new Promise((resolve, reject) => {
+      this.actions.push(async () => {
+        try {
+          resolve(await action());
+        } catch (ex) {
+          reject(ex);
+        }
+      });
+
+      if (this.actions.length === 1) {
+        this.pumpQueue();
+      }
+    });
+  }
+
```

Comment:
This whole thing is kind of tricky. Basically, we want one thing pumping the queue at a time. The way it works right now, we have a queue pumping only when we need to.

I.e. after we first enqueue we will fire the pump function and forget about it. It will keep going while anything adds to it during the time it's suspended. It will then eventually die, and the next thing to queue an item will restart it.